src/HydroShocktubeCMA/test_hydro_shocktube_cma.cpp

```diff
@@ -0,0 +1,285 @@
+
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 0) = 0.8;
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 1) = 0.3 * pow(sin(20 * 3.14 * 0), 2);
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 2) = 1 - 0.8 - 0.3 * pow(sin(20 * 3.14 * 0), 2);
+
+	} else if (i >= hi[0]) {
+		// x1 right-side boundary -- constant
+		for (int n = 0; n < numcomp; ++n) {
+			consVar(i, j, k, n) = 0;
+		}
+
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::gasEnergy_index) = P_R / (gamma - 1.);
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::gasInternalEnergy_index) = P_R / (gamma - 1.);
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::gasDensity_index) = rho_R;
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::x1GasMomentum_index) = 0.;
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::x2GasMomentum_index) = 0.;
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::x3GasMomentum_index) = 0.;
+
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 0) = 0.1;
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 1) = 0.3 * pow(sin(20 * 3.14 * 1), 2);
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 2) = 1 - 0.1 - 0.3 * pow(sin(20 * 3.14 * 1), 2);
+	}
+}
+
+template <> void RadhydroSimulation<ShocktubeProblem>::ErrorEst(int lev, amrex::TagBoxArray &tags, Real /*time*/, int /*ngrow*/)
+{
+	// tag cells for refinement
+
+	const Real eta_threshold = 0.1; // gradient refinement threshold
+	const Real rho_min = 0.01;	// minimum rho for refinement
+	auto const &dx = geom[lev].CellSizeArray();
+
+	for (amrex::MFIter mfi(state_new_cc_[lev]); mfi.isValid(); ++mfi) {
+		const amrex::Box &box = mfi.validbox();
+		const auto state = state_new_cc_[lev].const_array(mfi);
+		const auto tag = tags.array(mfi);
+
+		amrex::ParallelFor(box, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			int const n = 0;
+			Real const rho = state(i, j, k, n);
+			Real const del_x = (state(i + 1, j, k, n) - state(i - 1, j, k, n)) / (2.0 * dx[0]);
+			Real const gradient_indicator = std::sqrt(del_x * del_x) / rho;
+
+			if (gradient_indicator > eta_threshold && rho >= rho_min) {
+				tag(i, j, k) = amrex::TagBox::SET;
+			}
+		});
+	}
+}
+
+template <> void RadhydroSimulation<ShocktubeProblem>::computeAfterTimestep()
+{
+	auto [position, values] = fextract(state_new_cc_[0], Geom(0), 0, 0.5);
+	const int nmscalars = Physics_Traits<ShocktubeProblem>::numMassScalars;
+
+	if (amrex::ParallelDescriptor::IOProcessor()) {
+		amrex::Real specieSum = 0.0;
+
+		for (int n = 0; n < nmscalars; ++n) {
+			specieSum += values.at(HydroSystem<ShocktubeProblem>::scalar0_index + n)[0];
+		}
+
+		const amrex::Real Delta_eps_t = 1e0 - specieSum;
+		amrex::Print() << "Mass scalar conservation: Delta_eps_t = " << Delta_eps_t << "\n";
+		userData_.scalarSum_vec_.push_back(Delta_eps_t);
+	}
+}
+
+template <> void RadhydroSimulation<ShocktubeProblem>::computeAfterEvolve(amrex::Vector<amrex::Real> &initSumCons)
```

Comment:
Is `initSumCons` the right thing to be called here? In this function below, I use `initSumCons` to find the values of mass scalars and then find `1 - sumMassScalars`

Reply:
`initSumCons` is the sum of the conserved variables in the *initial conditions*. If you look at RadhydroSimulation.hpp, it is used to subtract the initial mass/energy/momentum from the final values.

Reply:
We don't need `computeafterEvolve` anymore so this is redundant now.